File: libs/rust/src/credentials/service.rs
//! # Credential service (the testable seam)
//!
//! **One-liner purpose**: The public, transport-agnostic interface over the vault —
//! `gg.credentials()` returns a [`CredentialService`]; [`DefaultCredentialService`] wraps a
//! [`LocalVault`] behind a lock and refreshes cross-process changes on read.
//!
//! ## Overview
//! [`Secret`] carries the decrypted value in a [`zeroize`]-ing buffer and **never logs it**
//! ([`Secret`]'s `Debug` redacts the bytes). [`SecretMeta`] is metadata only. Typed convenience
//! views (AWS creds, basic-auth, TLS bundle, Kafka SASL) are deferred to a later phase and will be
//! thin accessors over [`Secret`].

use std::collections::BTreeMap;
use std::sync::{Arc, Mutex};

use zeroize::Zeroizing;

use super::sync::SyncEngine;
use super::vault::{LocalVault, PutOptions};
use crate::Result;
use crate::error::EdgeCommonsError;

/// A decrypted secret value plus its metadata. The bytes are zeroized on drop and redacted from
/// `Debug`; do not log or serialize them.
#[derive(Clone)]
pub struct Secret {
    pub name: String,
    pub version: String,
    pub(crate) bytes: Zeroizing<Vec<u8>>,
    pub labels: BTreeMap<String, String>,
    pub created_ms: u64,
    pub source: String,
    pub content_type: String,
}

impl Secret {
    /// The raw secret bytes.
    pub fn bytes(&self) -> &[u8] {
        &self.bytes
    }

    /// The value as UTF-8 (errors if not valid UTF-8).
    pub fn as_str(&self) -> Result<&str> {
        std::str::from_utf8(&self.bytes)
            .map_err(|_| EdgeCommonsError::Credentials("secret is not valid UTF-8".into()))
    }

    /// The value parsed as JSON.
    pub fn as_json(&self) -> Result<serde_json::Value> {
        serde_json::from_slice(&self.bytes)
            .map_err(|e| EdgeCommonsError::Credentials(format!("secret is not JSON: {e}")))
    }
}

impl std::fmt::Debug for Secret {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        f.debug_struct("Secret")
            .field("name", &self.name)
            .field("version", &self.version)
            .field(
                "bytes",
                &format_args!("<{} bytes redacted>", self.bytes.len()),
            )
            .field("source", &self.source)
            .finish()
    }
}

/// Non-sensitive credential-subsystem stats (for the metrics bridge). Never includes values.
#[derive(Debug, Clone, Default)]
pub struct CredentialStats {
    pub secret_count: u64,
    /// Age of the last successful central sync, ms (None if no central sync / never synced).
    pub last_sync_age_ms: Option<u64>,
    pub sync_failures: u64,
    pub rotations: u64,
}

/// Metadata for a secret version — safe to log/list (no value).
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct SecretMeta {
    pub name: String,
    pub version: String,
    pub created_ms: u64,
    pub ttl_secs: Option<u64>,
    pub source: String,
    pub labels: BTreeMap<String, String>,
}

/// The public credential interface (depend on this, not [`DefaultCredentialService`]).
pub trait CredentialService: Send + Sync {
    /// Latest version of `name`, or `None`.
    fn get(&self, name: &str) -> Result<Option<Secret>>;
    /// A specific version of `name`.
    fn get_version(&self, name: &str, version: &str) -> Result<Option<Secret>>;
    /// Whether a secret exists.
    fn exists(&self, name: &str) -> Result<bool>;
    /// Metadata for all secrets under `prefix` (empty = all). Never returns values.
    fn list(&self, prefix: &str) -> Result<Vec<SecretMeta>>;
    /// Retained version ids for `name` (oldest→newest).
    fn versions(&self, name: &str) -> Result<Vec<String>>;
    /// Write a local-only secret version; returns the new version id.
    fn put(&self, name: &str, value: &[u8], opts: PutOptions) -> Result<String>;
    /// Remove a secret entirely.
    fn delete(&self, name: &str) -> Result<bool>;
    /// Force an immediate pull from the central source (no-op when no central sync is configured).
    fn refresh(&self) -> Result<()> {
        Ok(())
    }

    /// Non-sensitive stats for observability (default: just the secret count).
    fn stats(&self) -> CredentialStats {
        CredentialStats {
            secret_count: self.list("").map(|v| v.len() as u64).unwrap_or(0),
            ..CredentialStats::default()
        }
    }

    /// The value as bytes (convenience).
    fn get_bytes(&self, name: &str) -> Result<Option<Zeroizing<Vec<u8>>>> {
        Ok(self.get(name)?.map(|s| s.bytes.clone()))
    }
    /// The value as a UTF-8 string (convenience).
    fn get_string(&self, name: &str) -> Result<Option<String>> {
        match self.get(name)? {
            Some(s) => Ok(Some(s.as_str()?.to_string())),
            None => Ok(None),
        }
    }
    /// The value parsed as JSON (convenience).
    fn get_json(&self, name: &str) -> Result<Option<serde_json::Value>> {
        match self.get(name)? {
            Some(s) => Ok(Some(s.as_json()?)),
            None => Ok(None),
        }
    }

    // ----- typed views (thin parses over the opaque secret; see `views`) -----

    /// AWS credentials stored at `name`.
    fn get_aws_credentials(&self, name: &str) -> Result<Option<super::views::AwsCredentials>> {
        match self.get(name)? {
            Some(s) => Ok(Some(super::views::parse(&s, "AWS credentials")?)),
            None => Ok(None),
        }
    }
    /// HTTP basic-auth credentials stored at `name`.
    fn get_basic_auth(&self, name: &str) -> Result<Option<super::views::BasicAuth>> {
        match self.get(name)? {
            Some(s) => Ok(Some(super::views::parse(&s, "basic auth")?)),
            None => Ok(None),
        }
    }
    /// A TLS bundle stored at `name`.
    fn get_tls_bundle(&self, name: &str) -> Result<Option<super::views::TlsBundle>> {
        match self.get(name)? {
            Some(s) => Ok(Some(super::views::parse(&s, "a TLS bundle")?)),
            None => Ok(None),
        }
    }
    /// Kafka SASL credentials stored at `name`.
    fn get_kafka_sasl(&self, name: &str) -> Result<Option<super::views::KafkaSasl>> {
        match self.get(name)? {
            Some(s) => Ok(Some(super::views::parse(&s, "Kafka SASL")?)),
            None => Ok(None),
        }
    }
}

/// The default [`CredentialService`]: a [`LocalVault`] behind a mutex. Each read first picks up any
/// cross-process change (the shared device vault may be written by another component).
pub struct DefaultCredentialService {
    vault: Arc<Mutex<LocalVault>>,
    /// Owns the central sync background thread (RAII); `None` for a standalone local vault.
    _sync: Option<SyncEngine>,
    /// Transparent key namespace (`<thingName>/<componentName>`), or empty for no namespacing.
    /// Prepended to every key so a shared device vault (and a fleet's central store) can't collide
    /// across components/devices; stripped from returned names so callers see their own keys.
    namespace: String,
    /// Audit sink for access events (`None` = auditing off). Set via [`with_audit`](Self::with_audit);
    /// the config path enables it (`credentials.audit.enabled`) with the default logging sink.
    audit: Option<Arc<dyn super::audit::AuditSink>>,
}

impl DefaultCredentialService {
    /// Wrap an opened [`LocalVault`] (standalone, no central sync, no namespacing, no audit).
    pub fn new(vault: LocalVault) -> Self {
        Self {
            vault: Arc::new(Mutex::new(vault)),
            _sync: None,
            namespace: String::new(),
            audit: None,
        }
    }

    /// Wrap a shared vault that a [`SyncEngine`] also writes to, with the given key namespace.
    pub fn with_sync(
        vault: Arc<Mutex<LocalVault>>,
        sync: Option<SyncEngine>,
        namespace: String,
    ) -> Self {
        Self {
            vault,
            _sync: sync,
            namespace,
            audit: None,
        }
    }

    /// Attach (or clear) the audit sink — access events are emitted to it. Fluent; returns `self`.
    pub fn with_audit(mut self, sink: Option<Arc<dyn super::audit::AuditSink>>) -> Self {
        self.audit = sink;
        self
    }

    /// Emit an audit event if an audit sink is configured (no-op otherwise).
    fn audit_event(
        &self,
        op: &'static str,
        name: &str,
        version: &str,
        source: &str,
        outcome: &'static str,
    ) {
        if let Some(sink) = &self.audit {
            sink.record(&super::audit::AuditEvent {
                op,
                name: name.to_string(),
                version: version.to_string(),
                source: source.to_string(),
                outcome,
            });
        }
    }

    /// The shared vault handle (so a [`SyncEngine`] can be constructed against the same store).
    pub fn vault_arc(&self) -> Arc<Mutex<LocalVault>> {
        self.vault.clone()
    }

    fn locked(&self) -> std::sync::MutexGuard<'_, LocalVault> {
        self.vault.lock().unwrap_or_else(|p| p.into_inner())
    }

    /// Map a caller-facing key to its namespaced storage key.
    fn full(&self, name: &str) -> String {
        if self.namespace.is_empty() {
            name.to_string()
        } else {
            format!("{}/{}", self.namespace, name)
        }
    }

    /// Strip the namespace from a storage key for the caller.
    fn rel(&self, full: &str) -> String {
        if self.namespace.is_empty() {
            full.to_string()
        } else {
            full.strip_prefix(&format!("{}/", self.namespace))
                .unwrap_or(full)
                .to_string()
        }
    }
}

impl CredentialService for DefaultCredentialService {
    fn get(&self, name: &str) -> Result<Option<Secret>> {
        // Scope the vault lock so it's released before the audit sink is called.
        let result = {
            let mut v = self.locked();
            v.reload_if_changed()?;
            v.get(&self.full(name))?.map(|mut s| {
                s.name = self.rel(&s.name);
                s
            })
        };
        match &result {
            Some(s) => self.audit_event("get", name, &s.version, &s.source, "hit"),
            None => self.audit_event("get", name, "-", "-", "miss"),
        }
        Ok(result)
    }
    fn get_version(&self, name: &str, version: &str) -> Result<Option<Secret>> {
        let result = {
            let mut v = self.locked();
            v.reload_if_changed()?;
            v.get_version(&self.full(name), version)?.map(|mut s| {
                s.name = self.rel(&s.name);
                s
            })
        };
        match &result {
            Some(s) => self.audit_event("get", name, &s.version, &s.source, "hit"),
            None => self.audit_event("get", name, version, "-", "miss"),
        }
        Ok(result)
    }
    fn exists(&self, name: &str) -> Result<bool> {
        let mut v = self.locked();
        v.reload_if_changed()?;
        Ok(v.exists(&self.full(name)))
    }
    fn list(&self, prefix: &str) -> Result<Vec<SecretMeta>> {
        let mut v = self.locked();
        v.reload_if_changed()?;
        // List within this component's namespace and strip it from the returned names.
        Ok(v.list(&self.full(prefix))
            .into_iter()
            .map(|mut m| {
                m.name = self.rel(&m.name);
                m
            })
            .collect())
    }
    fn versions(&self, name: &str) -> Result<Vec<String>> {
        let mut v = self.locked();
        v.reload_if_changed()?;
        Ok(v.versions(&self.full(name)))
    }
    fn put(&self, name: &str, value: &[u8], opts: PutOptions) -> Result<String> {
        let version = {
            let mut v = self.locked();
            v.reload_if_changed()?;
            v.put(&self.full(name), value, opts)?
        };
        self.audit_event("put", name, &version, "local", "ok");
        Ok(version)
    }
    fn delete(&self, name: &str) -> Result<bool> {
        let deleted = {
            let mut v = self.locked();
            v.reload_if_changed()?;
            v.delete(&self.full(name))?
        };
        self.audit_event(
            "delete",
            name,
            "-",
            "-",
            if deleted { "ok" } else { "miss" },
        );
        Ok(deleted)
    }
    fn refresh(&self) -> Result<()> {
        if let Some(sync) = &self._sync {
            sync.sync_now();
        }
        Ok(())
    }

    fn stats(&self) -> CredentialStats {
        let secret_count = self.list("").map(|v| v.len() as u64).unwrap_or(0);
        let (last_sync_age_ms, sync_failures, rotations) = match &self._sync {
            Some(s) => {
                let (last_ok, failures, rotations) = s.stats();
                (
                    last_ok.map(|ms| now_ms_service().saturating_sub(ms)),
                    failures,
                    rotations,
                )
            }
            None => (None, 0, 0),
        };
        CredentialStats {
            secret_count,
            last_sync_age_ms,
            sync_failures,
            rotations,
        }
    }
}

fn now_ms_service() -> u64 {
    std::time::SystemTime::now()
        .duration_since(std::time::UNIX_EPOCH)
        .map(|d| d.as_millis() as u64)
        .unwrap_or(0)
}

File: libs/python/edgecommons/credentials/service.py
"""Credential service (the public seam) + Secret / SecretMeta value types."""
import json
import threading
import time
from dataclasses import dataclass, field, replace
from typing import Dict, List, Optional

from .errors import CredentialError
from .views import AwsCredentials, BasicAuth, KafkaSasl, TlsBundle, _require


class Secret:
    """A decrypted secret value plus metadata. ``repr`` redacts the bytes — never log the value."""

    def __init__(self, name: str, version: str, _bytes: bytes, labels: Dict[str, str],
                 created_ms: int, source: str, content_type: str):
        self.name = name
        self.version = version
        self._bytes = _bytes
        self.labels = labels
        self.created_ms = created_ms
        self.source = source
        self.content_type = content_type

    def bytes(self) -> bytes:
        return self._bytes

    def as_str(self) -> str:
        try:
            return self._bytes.decode("utf-8")
        except UnicodeDecodeError:
            raise CredentialError("secret is not valid UTF-8") from None

    def as_json(self):
        try:
            return json.loads(self._bytes)
        except ValueError as e:
            raise CredentialError(f"secret is not JSON: {e}") from None

    def __repr__(self) -> str:
        return f"Secret(name={self.name!r}, version={self.version!r}, bytes=<{len(self._bytes)} redacted>)"


@dataclass
class CredentialStats:
    """Non-sensitive credential-subsystem stats (for the metrics bridge). Never includes values."""
    secret_count: int = 0
    # Age of the last successful central sync, ms (None if no central sync / never synced).
    last_sync_age_ms: Optional[int] = None
    sync_failures: int = 0
    rotations: int = 0


@dataclass(frozen=True)
class SecretMeta:
    """Metadata for a secret version — safe to log/list (no value)."""
    name: str
    version: str
    created_ms: int
    source: str
    ttl_secs: Optional[int] = None
    labels: Dict[str, str] = field(default_factory=dict)


class CredentialService:
    """Public interface over the vault. Depend on this; the default impl wraps a LocalVault."""

    def get(self, name: str) -> Optional[Secret]:
        raise NotImplementedError

    def get_version(self, name: str, version: str) -> Optional[Secret]:
        raise NotImplementedError

    def exists(self, name: str) -> bool:
        raise NotImplementedError

    def list(self, prefix: str = "") -> List[SecretMeta]:
        raise NotImplementedError

    def versions(self, name: str) -> List[str]:
        raise NotImplementedError

    def put(self, name: str, value: bytes, **opts) -> str:
        raise NotImplementedError

    def delete(self, name: str) -> bool:
        raise NotImplementedError

    def refresh(self) -> None:
        """Force an immediate pull from the central source (no-op without central sync)."""
        return None

    def stats(self) -> CredentialStats:
        """Non-sensitive stats for observability (default: just the secret count)."""
        try:
            count = len(self.list(""))
        except Exception:
            count = 0
        return CredentialStats(secret_count=count)

    # convenience views
    def get_bytes(self, name: str) -> Optional[bytes]:
        s = self.get(name)
        return s.bytes() if s else None

    def get_string(self, name: str) -> Optional[str]:
        s = self.get(name)
        return s.as_str() if s else None

    def get_json(self, name: str):
        s = self.get(name)
        return s.as_json() if s else None

    # ----- typed views (thin parses over the opaque secret) -----
    def get_aws_credentials(self, name: str):
        s = self.get(name)
        if s is None:
            return None
        d = s.as_json()
        return AwsCredentials(_require(d, "accessKeyId", "AWS credentials"),
                              _require(d, "secretAccessKey", "AWS credentials"),
                              d.get("sessionToken"), d.get("expiry"))

    def get_basic_auth(self, name: str):
        s = self.get(name)
        if s is None:
            return None
        d = s.as_json()
        return BasicAuth(_require(d, "username", "basic auth"), _require(d, "password", "basic auth"))

    def get_tls_bundle(self, name: str):
        s = self.get(name)
        if s is None:
            return None
        d = s.as_json()
        return TlsBundle(_require(d, "certPem", "a TLS bundle"), _require(d, "keyPem", "a TLS bundle"), d.get("caPem"))

    def get_kafka_sasl(self, name: str):
        s = self.get(name)
        if s is None:
            return None
        d = s.as_json()
        return KafkaSasl(_require(d, "username", "Kafka SASL"), _require(d, "password", "Kafka SASL"),
                         d.get("mechanism", "PLAIN"))


class DefaultCredentialService(CredentialService):
    """A :class:`~edgecommons.credentials.vault.LocalVault` behind a lock; refreshes on read.

    ``namespace`` (``<thingName>/<componentName>``) is transparently prepended to every key and
    stripped from returned names, so a shared device vault can't collide across components.
    """

    def __init__(self, vault, namespace: str = "", sync=None, lock: Optional[threading.Lock] = None,
                 audit=None):
        self._vault = vault
        self._lock = lock if lock is not None else threading.Lock()
        self._namespace = namespace
        self._sync = sync
        # Audit sink for access events (None = auditing off). The config path enables it
        # (credentials.audit.enabled) with the default logging sink.
        self._audit = audit

    def with_audit(self, sink) -> "DefaultCredentialService":
        """Attach (or clear) the audit sink — access events are emitted to it. Fluent; returns self."""
        self._audit = sink
        return self

    def _emit_audit(self, op: str, name: str, version: str, source: str, outcome: str) -> None:
        """Emit an audit event if an audit sink is configured (no-op otherwise)."""
        if self._audit is not None:
            from .audit import AuditEvent
            self._audit.record(AuditEvent(op=op, name=name, version=version, source=source, outcome=outcome))

    def _full(self, name: str) -> str:
        return f"{self._namespace}/{name}" if self._namespace else name

    def _rel(self, full: str) -> str:
        prefix = self._namespace + "/"
        return full[len(prefix):] if self._namespace and full.startswith(prefix) else full

    def get(self, name: str) -> Optional[Secret]:
        with self._lock:
            self._vault.reload_if_changed()
            s = self._vault.get(self._full(name))
        if s is not None:
            s.name = self._rel(s.name)
            self._emit_audit("get", name, s.version, s.source, "hit")
        else:
            self._emit_audit("get", name, "-", "-", "miss")
        return s

    def get_version(self, name: str, version: str) -> Optional[Secret]:
        with self._lock:
            self._vault.reload_if_changed()
            s = self._vault.get_version(self._full(name), version)
        if s is not None:
            s.name = self._rel(s.name)
            self._emit_audit("get", name, s.version, s.source, "hit")
        else:
            self._emit_audit("get", name, version, "-", "miss")
        return s

    def exists(self, name: str) -> bool:
        with self._lock:
            self._vault.reload_if_changed()
            return self._vault.exists(self._full(name))

    def list(self, prefix: str = "") -> List[SecretMeta]:
        with self._lock:
            self._vault.reload_if_changed()
            metas = self._vault.list(self._full(prefix))
        return [replace(m, name=self._rel(m.name)) for m in metas]

    def versions(self, name: str) -> List[str]:
        with self._lock:
            self._vault.reload_if_changed()
            return self._vault.versions(self._full(name))

    def put(self, name: str, value: bytes, **opts) -> str:
        with self._lock:
            self._vault.reload_if_changed()
            version = self._vault.put(self._full(name), value, **opts)
        self._emit_audit("put", name, version, "local", "ok")
        return version

    def delete(self, name: str) -> bool:
        with self._lock:
            self._vault.reload_if_changed()
            deleted = self._vault.delete(self._full(name))
        self._emit_audit("delete", name, "-", "-", "ok" if deleted else "miss")
        return deleted

    def refresh(self) -> None:
        if self._sync is not None:
            self._sync.sync_now()

    def stats(self) -> CredentialStats:
        try:
            secret_count = len(self.list(""))
        except Exception:
            secret_count = 0
        last_sync_age_ms: Optional[int] = None
        sync_failures = 0
        rotations = 0
        if self._sync is not None:
            last_ok, sync_failures, rotations = self._sync.stats()
            if last_ok is not None:
                last_sync_age_ms = max(0, int(time.time() * 1000) - last_ok)
        return CredentialStats(
            secret_count=secret_count,
            last_sync_age_ms=last_sync_age_ms,
            sync_failures=sync_failures,
            rotations=rotations,
        )
